### PR TITLE
add fast imresize

### PR DESCRIPTION
**Pull request by JoannaLXY**
add fast imresize by using cv2 imresize with predefined output destination

### CODE REVIEW

File: mmcv/image/transforms/resize.py [hunk 1]
```diff
@@ -69,6 +101,31 @@ def imresize_like(img, dst_img, return_scale=False, interpolation='bilinear'):
     return imresize(img, (w, h), return_scale, interpolation)
 
 
+def imrescale_size(old_size, scale, return_scale=False):
+    w, h = old_size[:]
```
Comment (innerlee) [inline mmcv/image/transforms/resize.py]: `w, h = old_size`

File: mmcv/image/transforms/resize.py [hunk 2]
```diff
@@ -27,12 +27,17 @@ def _scale_size(size, scale):
...
+             size,
+             return_scale=False,
+             interpolation='bilinear',
+             out=None):
     """Resize image to a given size.
 
     Args:
         img (ndarray): The input image.
         size (tuple): Target (w, h).
+        out (ndarray): The output destination.
```
Comment (innerlee) [inline mmcv/image/transforms/resize.py]: move to below

File: tests/test_image.py [hunk 3]
```diff
@@ -258,6 +258,10 @@ def test_imresize(self):
                                                       True)
         assert (resized_img.shape == (600, 1000, 3) and w_scale == 2.5
                 and h_scale == 2.0)
+        resized_img_dst = np.empty((600, 1000, 3), dtype=self.img.dtype)
+        resized_img = mmcv.imresize(self.img, (1000, 600), out=resized_img_dst)
+        assert id(resized_img_dst) == id(resized_img)
+        assert_array_equal(resized_img_dst, resized_img)
```
Comment (innerlee) [inline tests/test_image.py]:
```suggestion
        assert_array_equal(resized_img_dst, mmcv.imresize(self.img, (1000, 600)))
```

**Comment by hellock** [comment]
The two versions are almost the same. The new version is even a little bit slower.

```python
import mmcv
import numpy as np

np.random.seed(0)
img_large = np.random.random((1000, 800, 3))
img_small = np.random.random((300, 400, 3))

mid_size = (560, 570)

num = 100

with mmcv.Timer(print_tmpl='large to medium [old]: {:.5f}'):
    for _ in range(num):
        mmcv.imresize(img_large, mid_size)

with mmcv.Timer(print_tmpl='large to medium [new]: {:.5f}'):
    for _ in range(num):
        dst = np.empty_like(img_large)
        mmcv.imresize(img_large, mid_size, out=dst)

with mmcv.Timer(print_tmpl='small to medium [old]: {:.5f}'):
    for _ in range(num):
        mmcv.imresize(img_small, mid_size)

with mmcv.Timer(print_tmpl='small to medium [new]: {:.5f}'):
    for _ in range(num):
        dst = np.empty_like(img_small)
        mmcv.imresize(img_small, mid_size, out=dst)
```

Results:
```
large to medium [old]: 0.09417
large to medium [new]: 0.11374
small to medium [old]: 0.04599
small to medium [new]: 0.04774
```

**Comment by innerlee** [comment]
This is used for video processing, where a batch of images (of same size) need to be resized (into another size). The speed up originates from less memory allocation.

See this benchmark:
https://github.com/innerlee/cvbenchmark/blob/master/benchmarks/run_resize.py

**Comment by hellock** [comment]
> This is used for video processing, where a batch of images (of same size) need to be resized (into another size). The speed up originates from less memory allocation.
> 
> See this benchmark:
> https://github.com/innerlee/cvbenchmark/blob/master/benchmarks/run_resize.py

I suggest attaching a benchmark with each PR for speed optimization.

**Comment by innerlee** [comment]
good suggestion

**Comment by JoannaLXY** [comment]
For benchmark:
```python
import mmcv
import numpy as np

np.random.seed(0)
imgs_large = np.array([np.random.random((1000, 800, 3)) for _ in range(32)])
imgs_small = np.array([np.random.random((300, 400, 3)) for _ in range(32)])

mid_size = (560, 570)

num = 100

with mmcv.Timer(print_tmpl='imrescale: large to medium [old]: {:.5f}'):
    for _ in range(num):
        tuple_list = [
            mmcv.imrescale(img, mid_size, return_scale=True)
            for img in imgs_large
        ]
        imgs, scale_factors = list(zip(*tuple_list))
        scale_factor = scale_factors[0]
        imgs = np.array(imgs)

with mmcv.Timer(print_tmpl='imresize: large to medium [old]: {:.5f}'):
    for _ in range(num):
        tuple_list = [
            mmcv.imresize(img, mid_size, return_scale=True)
            for img in imgs_large
        ]
        imgs, w_scales, h_scales = list(zip(*tuple_list))
        scale_factor = np.array(
            [w_scales[0], h_scales[0], w_scales[0], h_scales[0]],
            dtype=np.float32)
        imgs = np.array(imgs)

with mmcv.Timer(print_tmpl='imrescale: large to medium [new]: {:.5f}'):
    for _ in range(num):
        n, oh, ow, c = imgs_large.shape
        new_size, scale_factor = mmcv.image.imrescale_size((ow, oh),
                                                        mid_size,
                                                        return_scale=True)
        w, h = new_size

        rimgs = np.empty((n, h, w, c), dtype=imgs.dtype)
        for i in range(n):
            mmcv.imresize(imgs_large[i], (w, h), out=rimgs[i])

with mmcv.Timer(print_tmpl='imresize: large to medium [new]: {:.5f}'):
    for _ in range(num):
        n, oh, ow, c = imgs_large.shape
        w, h = mid_size
        scale_factor = np.array([w / ow, h / oh, w / ow, h / oh],
                                dtype=np.float32)

        rimgs = np.empty((n, h, w, c), dtype=imgs.dtype)
        for i in range(n):
            mmcv.imresize(imgs_large[i], (w, h), out=rimgs[i])

with mmcv.Timer(print_tmpl='imrescale: small to medium [old]: {:.5f}'):
    for _ in range(num):
        tuple_list = [
            mmcv.imrescale(img, mid_size, return_scale=True)
            for img in imgs_small
        ]
        imgs, scale_factors = list(zip(*tuple_list))
        scale_factor = scale_factors[0]
        imgs = np.array(imgs)

with mmcv.Timer(print_tmpl='imresize: small to medium [old]: {:.5f}'):
    for _ in range(num):
        tuple_list = [
            mmcv.imresize(img, mid_size, return_scale=True)
            for img in imgs_small
        ]
        imgs, w_scales, h_scales = list(zip(*tuple_list))
        scale_factor = np.array(
            [w_scales[0], h_scales[0], w_scales[0], h_scales[0]],
            dtype=np.float32)
        imgs = np.array(imgs)

with mmcv.Timer(print_tmpl='imrescale: small to medium [new]: {:.5f}'):
    for _ in range(num):
        n, oh, ow, c = imgs_small.shape
        new_size, scale_factor = mmcv.image.imrescale_size((ow, oh),
                                                        mid_size,
                                                        return_scale=True)
        w, h = new_size

    rimgs = np.empty((n, h, w, c), dtype=imgs.dtype)
    for i in range(n):
        mmcv.imresize(imgs_small[i], (w, h), out=rimgs[i])

with mmcv.Timer(print_tmpl='imresize: large to medium [new]: {:.5f}'):
    for _ in range(num):
        n, oh, ow, c = imgs_small.shape
        w, h = mid_size
        scale_factor = np.array([w / ow, h / oh, w / ow, h / oh],
                                dtype=np.float32)

        rimgs = np.empty((n, h, w, c), dtype=imgs.dtype)
        for i in range(n):
            mmcv.imresize(imgs_small[i], (w, h), out=rimgs[i])

```

Results:
```
imrescale: large to medium [old]: 13.40840
imresize: large to medium [old]: 15.45227
imrescale: large to medium [new]: 5.92301
imresize: large to medium [new]: 5.66638
imrescale: small to medium [old]: 11.76296
imresize: small to medium [old]: 14.38052
imrescale: small to medium [new]: 4.50699
imresize: large to medium [new]: 4.18508
```

File: mmcv/image/transforms/resize.py [hunk 4]
```diff
@@ -69,6 +74,31 @@ def imresize_like(img, dst_img, return_scale=False, interpolation='bilinear'):
     return imresize(img, (w, h), return_scale, interpolation)
 
 
+def rescale_size(old_size, scale, return_scale=False):
```
Comment (hellock) [inline mmcv/image/transforms/resize.py]: Add docstring here.

**Pull request merged**